File: scripts/validate-repository-docs.py
#!/usr/bin/env python3
"""Validate README links, visual assets, evidence, version claims and repository hygiene."""

from __future__ import annotations

import json
import re
import subprocess
from pathlib import Path
import xml.etree.ElementTree as ET

ROOT = Path(__file__).resolve().parents[1]
README = ROOT / "README.md"
PACKAGE = ROOT / "package.json"
VALIDATION = ROOT / "docs" / "VALIDATION.md"
VISUAL_DESIGN_SYSTEM = ROOT / "docs" / "README_VISUAL_DESIGN_SYSTEM.md"
ASSETS = ROOT / "docs" / "assets"
LATEST_REPORT = ROOT / "reports" / "final-visual-upgrade-20260726" / "REPORT.md"
LATEST_SUMMARY = ROOT / "reports" / "final-visual-upgrade-20260726" / "summary.json"
LATEST_ALIAS = ROOT / "reports" / "ci-validation-latest.json"

EXPECTED_VISUALS = (
    "resindb-ai-platform-overview.svg",
    "resindb-data-lifecycle.svg",
    "resindb-data-governance.svg",
    "resindb-scientific-engine.svg",
    "resindb-worker-architecture.svg",
    "resindb-formula-engine.svg",
    "resindb-knowledge-network.svg",
    "resindb-comparison-decision.svg",
    "resindb-ai-workflow.svg",
    "resindb-local-first-privacy.svg",
    "resindb-import-export.svg",
    "resindb-quality-gates.svg",
    "resindb-research-workflow.svg",
    "resindb-security-deployment.svg",
)

FORBIDDEN_PATHS = (
    ".github/.resindb-main-update",
    ".github/.resindb-v310-patch",
    ".github/final-reaudit-20260726.trigger",
    ".github/source-hygiene-finalization-20260726.trigger",
    ".github/workflows/apply-resindb-main-update.yml",
    ".github/workflows/apply-v310-patch.yml",
    ".github/workflows/diagnose-v310-patch.yml",
    ".github/workflows/finalize-resindb-main-update.yml",
    ".github/workflows/repository-audit-20260726.yml",
    ".github/workflows/finalize-repository-20260726.yml",
    ".github/workflows/finalize-repository-v2-20260726.yml",
    ".github/workflows/final-proof-20260726.yml",
    ".github/workflows/final-proof-docs-20260726.yml",
    ".github/workflows/reaudit-20260726.yml",
    ".github/workflows/final-reaudit-20260726.yml",
    ".github/workflows/source-hygiene-finalization-20260726.yml",
    ".github/workflows/ultimate-readme-audit-20260726.yml",
    ".github/ultimate-readme-audit-20260726.trigger",
    ".github/final-dependency-remediation-20260726.trigger",
    ".github/workflows/final-dependency-remediation-20260726.yml",
    ".github/dependency-audit-diagnostic-20260726.trigger",
    ".github/workflows/dependency-audit-diagnostic-20260726.yml",
    "scripts/finalize-dependency-candidate.py",
    "reports/dependency-audit-diagnostic-20260726.json",
    "reports/dependency-audit-diagnostic-20260726.md",
    ".github/final-eslint-remediation-20260726.trigger",
    ".github/workflows/final-eslint-remediation-20260726.yml",
    ".github/current-audit-diagnostic-20260726.trigger",
    ".github/workflows/current-audit-diagnostic-20260726.yml",
    "reports/current-audit-diagnostic-20260726.json",
    ".github/final-compatible-lint-stack-20260726.trigger",
    ".github/workflows/final-compatible-lint-stack-20260726.yml",
    ".github/npm-ci-peer-diagnostic-20260726.trigger",
    ".github/workflows/npm-ci-peer-diagnostic-20260726.yml",
    "reports/npm-ci-peer-diagnostic-20260726.txt",
    ".github/final-lockfile-rebuild-20260726.trigger",
    ".github/workflows/final-lockfile-rebuild-20260726.yml",
    ".github/lint-stack-install-diagnostic-20260726.trigger",
    ".github/workflows/lint-stack-install-diagnostic-20260726.yml",
    "scripts/finalize-lockfile-rebuild.sh",
    "reports/lint-stack-install-diagnostic-20260726.txt",
    ".github/final-current-tree-proof-20260726.trigger",
    ".github/workflows/final-current-tree-proof-20260726.yml",
    "scripts/final-current-tree-proof.sh",
    ".github/fresh-lockfile-rebuild-20260726.trigger",
    ".github/workflows/fresh-lockfile-rebuild-20260726.yml",
    ".github/lockfile-rebuild-diagnostic-20260726.trigger",
    ".github/workflows/lockfile-rebuild-diagnostic-20260726.yml",
    "reports/lockfile-rebuild-diagnostic-20260726.txt",
    "reports/fresh-lockfile-rebuild-failure-20260726.md",
    ".github/uiux-design-system-research-20260727.trigger",
    ".github/workflows/uiux-design-system-research-20260727.yml",
    ".github/uiux-readme-visual-redesign-20260727.trigger",
    ".github/workflows/uiux-readme-visual-redesign-20260727.yml",
    "scripts/apply-uiux-readme-visual-system.py",
    "docs/MIGRATION_v3.1.0.md",
    "docs/RELEASE_NOTES_v3.1.0.md",
    "reports/patch-diagnostic.json",
    "reports/release-baseline-v3.1.0.json",
    "reports/automated-audit-20260726",
    "reports/final-reaudit-20260726",
)

LOCAL_LINK_PATTERNS = (
    re.compile(r'<(?:img|a)\b[^>]*(?:src|href)="([^"]+)"', re.IGNORECASE),
    re.compile(r'!?\[[^\]]*]\(([^)]+)\)'),
)


def fail(message: str) -> None:
    raise SystemExit(message)


def strip_target(raw: str) -> str:
    target = raw.strip().split(maxsplit=1)[0].strip("<>")
    target = target.split("#", 1)[0].split("?", 1)[0]
    return target


def validate_local_links(readme_text: str) -> None:
    missing: list[str] = []
    seen: set[str] = set()
    for pattern in LOCAL_LINK_PATTERNS:
        for raw in pattern.findall(readme_text):
            target = strip_target(raw)
            if not target or target.startswith(("http://", "https://", "mailto:", "data:", "#")):
                continue
            if target in seen:
                continue
            seen.add(target)
            if not (ROOT / target).exists():
                missing.append(target)
    if missing:
        fail(f"README contains missing local targets: {sorted(missing)}")


def validate_visual_inventory(readme_text: str) -> None:
    expected = set(EXPECTED_VISUALS)
    actual = {path.name for path in ASSETS.glob("resindb-*.svg")}
    if actual != expected:
        fail(
            "README visual inventory mismatch: "
            f"missing={sorted(expected - actual)}, extra={sorted(actual - expected)}"
        )
    namespace = "{http://www.w3.org/2000/svg}"
    for filename in EXPECTED_VISUALS:
        relative = f"docs/assets/{filename}"
        if readme_text.count(relative) != 1:
            fail(f"{relative} must be referenced exactly once in README")
        root = ET.parse(ASSETS / filename).getroot()
        if root.attrib.get("role") != "img" or not root.attrib.get("aria-labelledby"):
            fail(f"{filename}: missing role=img or aria-labelledby")
        if not root.findall(f"{namespace}title") or not root.findall(f"{namespace}desc"):
            fail(f"{filename}: missing title or desc")
        if not root.findall(f"{namespace}metadata"):
            fail(f"{filename}: missing design-system metadata")
        if root.attrib.get("data-design-system") != "resindb-uiux-pro-max-v1":
            fail(f"{filename}: missing ResinDB UI/UX Pro Max design-system marker")
        if filename == "resindb-quality-gates.svg":
            svg_text = (ASSETS / filename).read_text(encoding="utf-8")
            for phrase in ("Source hygiene", "production code • no injection"):
                if phrase not in svg_text:
                    fail(f"{filename}: quality-gate semantics are missing {phrase!r}")


def validate_visual_design_system(readme_text: str) -> None:
    if not VISUAL_DESIGN_SYSTEM.is_file():
        fail("README visual design-system document is missing")
    if "docs/README_VISUAL_DESIGN_SYSTEM.md" not in readme_text:
        fail("README must link the visual design-system document")
    design_text = VISUAL_DESIGN_SYSTEM.read_text(encoding="utf-8")
    required = (
        "UI/UX Pro Max",
        "Swiss Modernism 2.0",
        "Bento Grid",
        "Accessible & Ethical",
        "8-point",
        "data-design-system=\"resindb-uiux-pro-max-v1\"",
    )
    missing = [value for value in required if value not in design_text]
    if missing:
        fail(f"README visual design system is incomplete: {missing}")


def validate_version_and_scripts(readme_text: str, validation_text: str) -> None:
    package = json.loads(PACKAGE.read_text(encoding="utf-8"))
    version = package.get("version")
    scripts = package.get("scripts", {})
    required_scripts = {
        "visuals:generate": "python3 scripts/generate-readme-visuals.py",
        "visuals:check": "python3 scripts/generate-readme-visuals.py --check",
        "validate:docs": "python3 scripts/validate-repository-docs.py",
        "validate:source": "python3 scripts/validate-source-hygiene.py",
        "audit:all": "npm audit --audit-level=high",
        "audit:prod": "npm audit --omit=dev --audit-level=high",
    }
    for name, command in required_scripts.items():
        if scripts.get(name) != command:
            fail(f"package.json script {name!r} must equal {command!r}")
    validate_script = scripts.get("validate", "")
    if "npm run validate:docs" not in validate_script:
        fail("package.json validate must include npm run validate:docs")
    if "npm run validate:source" not in validate_script:
        fail("package.json validate must include npm run validate:source")
    if "npm run audit:all" not in scripts.get("validate:ci", ""):
        fail("package.json validate:ci must include npm run audit:all")
    if f"version-{version}-" not in readme_text:
        fail(f"README version badge is not aligned with package version {version}")
    if f"`{version}`" not in validation_text:
        fail(f"docs/VALIDATION.md does not identify package version {version}")
    expected_dev_dependencies = {
        "eslint": "^10.8.0",
        "@typescript-eslint/eslint-plugin": "^8.65.0",
        "@typescript-eslint/parser": "^8.65.0",
        "typescript-eslint": "^8.65.0",
        "eslint-plugin-react-hooks": "^7.1.1",
        "eslint-plugin-react-refresh": "^0.5.3",
    }
    dev_dependencies = package.get("devDependencies", {})
    for name, expected in expected_dev_dependencies.items():
        if dev_dependencies.get(name) != expected:
            fail(f"development toolchain drift: {name} must equal {expected}")
    if "十四张" not in readme_text and "14 张" not in readme_text:
        fail("README must state that the visual system contains 14 diagrams")


def validate_status_block(name: str, proof: object) -> None:
    if not isinstance(proof, dict):
        fail(f"{name} evidence is missing or malformed")
    if proof.get("result") != "success":
        fail(f"{name} result is not success")
    if proof.get("remoteBranches") != ["main"]:
        fail(f"{name} does not prove main is the sole remote branch")
    statuses = proof.get("statuses")
    if not isinstance(statuses, dict) or not statuses:
        fail(f"{name} contains no validation statuses")
    if any(not isinstance(code, int) or code != 0 for code in statuses.values()):
        fail(f"{name} contains a non-zero validation status")


def validate_latest_evidence(readme_text: str, validation_text: str) -> None:
    expected_paths = (
        "reports/final-visual-upgrade-20260726/REPORT.md",
        "reports/final-visual-upgrade-20260726/summary.json",
        "reports/ci-validation-latest.json",
    )
    for relative in expected_paths:
        if not (ROOT / relative).is_file():
            fail(f"latest durable validation evidence is missing: {relative}")
    if expected_paths[0] not in readme_text or expected_paths[2] not in readme_text:
        fail("README must link the current report and machine-readable alias")
    for relative in expected_paths:
        if relative not in validation_text:
            fail(f"docs/VALIDATION.md must identify current evidence: {relative}")
    for stale_prefix in (
        "reports/final-validation-20260726/",
        "reports/final-reaudit-20260726/",
    ):
        if stale_prefix in readme_text or stale_prefix in validation_text:
            fail(f"README or validation contract still points at superseded evidence: {stale_prefix}")

    summary = json.loads(LATEST_SUMMARY.read_text(encoding="utf-8"))
    alias = json.loads(LATEST_ALIAS.read_text(encoding="utf-8"))
    if summary != alias:
        fail("latest evidence alias differs from the fixed durable summary")
    validate_status_block("baseline", summary)
    current_tree = summary.get("currentTreeVerification")
    if current_tree is not None:
        validate_status_block("currentTreeVerification", current_tree)
    if summary.get("visualCount") != len(EXPECTED_VISUALS):
        fail("latest durable validation evidence does not cover all visuals")
    if sorted(summary.get("generatedVisuals", [])) != sorted(EXPECTED_VISUALS):
        fail("latest durable validation evidence visual inventory is incomplete")


def validate_ci() -> None:
    ci = (ROOT / ".github" / "workflows" / "ci.yml").read_text(encoding="utf-8")
    if "npm run validate:docs" not in ci:
        fail("permanent CI must execute npm run validate:docs")
    if "npm run validate:source" not in ci:
        fail("permanent CI must execute npm run validate:source")
    if "npm run audit:all" not in ci:
        fail("permanent CI must execute npm run audit:all")
    if "contents: write" in ci or "EMBEDDED DEPENDENCY REMEDIATION" in ci:
        fail("permanent CI must remain read-only and non-self-modifying")
    if "branches: [main]" not in ci:
        fail("permanent CI is not restricted to main")
    if "contents: read" not in ci:
        fail("permanent CI must retain read-only contents permission")


def validate_hygiene() -> None:
    present = [path for path in FORBIDDEN_PATHS if (ROOT / path).exists()]
    if present:
        fail(f"forbidden migration/diagnostic residue remains: {present}")


def main() -> None:
    readme_text = README.read_text(encoding="utf-8")
    validation_text = VALIDATION.read_text(encoding="utf-8")
    validate_local_links(readme_text)
    validate_visual_inventory(readme_text)
    validate_visual_design_system(readme_text)
    validate_version_and_scripts(readme_text, validation_text)
    validate_latest_evidence(readme_text, validation_text)
    validate_ci()
    validate_hygiene()
    subprocess.run(
        ["python3", "scripts/generate-readme-visuals.py", "--check"],
        cwd=ROOT,
        check=True,
    )
    print(
        f"validated README, {len(EXPECTED_VISUALS)} deterministic visuals, "
        "version/scripts, UI/UX Pro Max visual system, source hygiene contract, durable evidence, CI contract and repository hygiene"
    )


if __name__ == "__main__":
    main()

File: scripts/validate-repository-docs.mjs
#!/usr/bin/env node

import { createHash } from 'node:crypto'
import { existsSync, readFileSync, readdirSync } from 'node:fs'
import { dirname, relative, resolve, sep } from 'node:path'
import { fileURLToPath } from 'node:url'

const ROOT = resolve(dirname(fileURLToPath(import.meta.url)), '..')
const README_PATH = resolve(ROOT, 'README.md')
const PACKAGE_PATH = resolve(ROOT, 'package.json')
const VALIDATION_PATH = resolve(ROOT, 'docs/VALIDATION.md')
const IMAGE_DIR = resolve(ROOT, 'docs/images')
const DESIGN_PATH = resolve(ROOT, 'docs/README_VISUAL_DESIGN_SYSTEM.md')
const BUNDLE_PATH = resolve(ROOT, 'scripts/readme-visuals.bundle.json')
const EXPECTED = [
  'resindb-ai-platform-overview.svg', 'resindb-data-lifecycle.svg',
  'resindb-data-governance.svg', 'resindb-scientific-engine.svg',
  'resindb-worker-architecture.svg', 'resindb-formula-engine.svg',
  'resindb-structure-property.svg', 'resindb-molecular-analysis.svg',
  'resindb-rheology-model.svg', 'resindb-thermal-analysis.svg',
  'resindb-mechanical-analysis.svg', 'resindb-electrical-analysis.svg',
  'resindb-multiscale-simulation.svg', 'resindb-ai-material-discovery.svg',
  'resindb-ai-workflow.svg', 'resindb-knowledge-network.svg',
  'resindb-comparison-decision.svg', 'resindb-local-first-privacy.svg',
  'resindb-import-export.svg', 'resindb-security-deployment.svg',
  'resindb-research-workflow.svg', 'resindb-quality-gates.svg',
].sort()
const FORBIDDEN = [
  '.github/apply-final-v2.trigger', '.github/workflows/apply-final-v2.yml',
  '.github/apply-final-small.trigger', '.github/workflows/apply-final-small.yml',
  '.github/apply-final-optimization.trigger', '.github/workflows/apply-final-optimization.yml',
  '.github/diagnose-final-blobs.trigger', '.github/workflows/diagnose-final-blobs.yml',
  '.github/diagnose-original-head.trigger', '.github/workflows/diagnose-original-head.yml',
  '.github/final-source-export.trigger', '.github/workflows/final-source-export.yml',
  '.github/remediate-lockfile-20260726.trigger', '.github/workflows/remediate-lockfile-20260726.yml',
  'reports/_final_source_export', 'reports/patch-diagnostic.json', '.resindb-delete-manifest.txt',
  '.github/.resindb-final-patch', '.github/.resindb-v320-delta',
  '.github/.resindb-stage1-delta',
]
const FORBIDDEN_PATH_PATTERNS = [
  /^\.github\/\.resindb-[^/]*(?:delta|patch|transport)[^/]*(?:\/|$)/i,
]

function fail(message) { throw new Error(message) }
function sha256(content) { return createHash('sha256').update(content).digest('hex') }

function listRepositoryPaths(relativeRoot) {
  const absoluteRoot = resolve(ROOT, relativeRoot)
  if (!existsSync(absoluteRoot)) return []
  const paths = []
  const visit = (directory) => {
    for (const entry of readdirSync(directory, { withFileTypes: true })) {
      const absolutePath = resolve(directory, entry.name)
      const repositoryPath = relative(ROOT, absolutePath).split(sep).join('/')
      paths.push(repositoryPath)
      if (entry.isDirectory()) visit(absolutePath)
    }
  }
  visit(absoluteRoot)
  return paths
}

function validateLinks(text) {
  const patterns = [/<(?:img|a)\b[^>]*(?:src|href)="([^"]+)"/gi, /!?\[[^\]]*]\(([^)]+)\)/g]
  const missing = []
  const seen = new Set()
  for (const pattern of patterns) {
    for (const match of text.matchAll(pattern)) {
      const raw = match[1]
      const target = raw.trim().split(/\s+/, 1)[0].replace(/^<|>$/g, '').split('#', 1)[0].split('?', 1)[0]
      if (!target || /^(https?:|mailto:|data:|#)/.test(target) || seen.has(target)) continue
      seen.add(target)
      if (!existsSync(resolve(ROOT, target))) missing.push(target)
    }
  }
  if (missing.length) fail(`README contains missing local targets: ${JSON.stringify(missing.sort())}`)
}

function validateSvg(content, name) {
  const required = [
    /<svg\b[^>]*\brole="img"/,
    /<svg\b[^>]*\baria-labelledby="[^"]+"/,
    /<svg\b[^>]*\bdata-design-system="resindb-uiux-pro-max-v1"/,
    /<title\b[^>]*>[^<]+<\/title>/,
    /<desc\b[^>]*>[^<]+<\/desc>/,
    /<metadata>[^<]+<\/metadata>/,
  ]
  for (const pattern of required) if (!pattern.test(content)) fail(`${name}: accessibility/design metadata missing`)
}

const readme = readFileSync(README_PATH, 'utf8')
const packageJson = JSON.parse(readFileSync(PACKAGE_PATH, 'utf8'))
const validation = readFileSync(VALIDATION_PATH, 'utf8')
const version = packageJson.version
if (version !== '3.2.0' || !readme.includes(`version-${version}-`) || !validation.includes(`\`${version}\``)) {
  fail('Version drift between package, README and validation contract')
}
validateLinks(readme)
const actual = readdirSync(IMAGE_DIR).filter((name) => /^resindb-.*\.svg$/.test(name)).sort()
if (JSON.stringify(actual) !== JSON.stringify(EXPECTED)) fail('Visual inventory mismatch')
const bundle = JSON.parse(readFileSync(BUNDLE_PATH, 'utf8'))
if (bundle.fileCount !== 22 || Object.keys(bundle.files ?? {}).length !== 22) fail('README visual bundle must contain 22 assets')
for (const name of EXPECTED) {
  const relativePath = `docs/images/${name}`
  if (readme.split(relativePath).length - 1 !== 1) fail(`${relativePath} must be referenced exactly once`)
  const content = readFileSync(resolve(IMAGE_DIR, name))
  validateSvg(content.toString('utf8'), name)
  const metadata = bundle.files[name]
  if (!metadata || metadata.size !== content.length || metadata.sha256 !== sha256(content)) {
    fail(`${name}: bundle metadata differs from committed SVG`)
  }
}
if (!readme.includes('22 张')) fail('README must declare the 22-diagram scientific visual system')
if (!existsSync(DESIGN_PATH) || !readFileSync(DESIGN_PATH, 'utf8').includes('UI/UX Pro Max')) fail('Visual design system document missing')
const requiredScripts = {
  'visuals:bundle': 'node scripts/bundle-readme-visuals.mjs',
  'visuals:generate': 'node scripts/generate-readme-visuals.mjs',
  'visuals:check': 'node scripts/generate-readme-visuals.mjs --check',
  'validate:docs': 'node scripts/validate-repository-docs.mjs',
  'validate:source': 'node scripts/validate-source-hygiene.mjs',
  'audit:all': 'npm audit --audit-level=high',
  'report:pdf': 'node scripts/generate-validation-pdf.mjs',
}
for (const [key, value] of Object.entries(requiredScripts)) {
  if (packageJson.scripts?.[key] !== value) fail(`package script drift: ${key}`)
}
for (const path of ['data/manifest.json', 'data/version.json', 'data/metadata.json', 'data/resins/manifest.json']) {
  if (!existsSync(resolve(ROOT, path))) fail('Governed root data contract is incomplete')
}
const discoveredPaths = [
  ...listRepositoryPaths('.github'),
  ...listRepositoryPaths('reports'),
]
const residue = [...new Set([
  ...FORBIDDEN.filter((path) => existsSync(resolve(ROOT, path))),
  ...discoveredPaths.filter((path) => FORBIDDEN_PATH_PATTERNS.some((pattern) => pattern.test(path))),
])].sort()
if (residue.length) fail(`Temporary migration/diagnostic residue remains: ${JSON.stringify(residue)}`)
const pythonFiles = readdirSync(resolve(ROOT, 'scripts')).filter((name) => name.endsWith('.py'))
if (pythonFiles.length) fail(`Stage-one Node-only tooling is incomplete: ${JSON.stringify(pythonFiles)}`)
const ci = readFileSync(resolve(ROOT, '.github/workflows/ci.yml'), 'utf8')
for (const phrase of ['branches: [main]', 'contents: read', 'npm run validate:docs', 'npm run validate:source', 'npm audit --omit=dev --audit-level=high --json', 'npm audit --audit-level=high --json', 'npm run report:pdf']) {
  if (!ci.includes(phrase)) fail(`Permanent CI missing: ${phrase}`)
}
if (ci.includes('contents: write')) fail('Permanent CI must be read-only')
if (/\bpython(?:3)?\b/i.test(ci)) fail('Permanent CI must not depend on Python')
console.log(`Validated README links, ${EXPECTED.length} deterministic scientific visuals, version ${version}, Node-only tooling, governed data contract and repository hygiene.`)
